File: mcp/mcp_server.mjs
/**
 * MCP Server for GraphQL Bookstore API
 * Allows AI/LLM models to interact with the Bookstore GraphQL API via MCP protocol.
 * 
 * Usage:
 *   npm install
 *   npm start
 * 
 * Or with custom API URL:
 *   API_URL=http://localhost:4000/graphql npm start
 */

import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const { Server } = await import(join(__dirname, 'node_modules/@modelcontextprotocol/sdk/dist/server/index.js'));
const { StdioServerTransport } = await import(join(__dirname, 'node_modules/@modelcontextprotocol/sdk/dist/server/stdio.js'));
const { CallToolRequestSchema } = await import(join(__dirname, 'node_modules/@modelcontextprotocol/sdk/dist/types.js'));
const { ListResourcesRequestSchema } = await import(join(__dirname, 'node_modules/@modelcontextprotocol/sdk/dist/types.js'));
const { ListToolsRequestSchema } = await import(join(__dirname, 'node_modules/@modelcontextprotocol/sdk/dist/types.js'));
const { ReadResourceRequestSchema } = await import(join(__dirname, 'node_modules/@modelcontextprotocol/sdk/dist/types.js'));

const API_URL = process.env.API_URL || 'http://127.0.0.1:4000/graphql';
const DEFAULT_USERNAME = 'admin';
const DEFAULT_PASSWORD = 'password123';

class BookstoreMCPClient {
  constructor() {
    this.token = null;
  }

  async _execute(query, variables = {}) {
    const headers = { 'Content-Type': 'application/json' };
    if (this.token) {
      headers['Authorization'] = `Bearer ${this.token}`;
    }

    const response = await fetch(API_URL, {
      method: 'POST',
      headers,
      body: JSON.stringify({ query, variables }),
    });

    if (!response.ok) {
      throw new Error(`HTTP error: ${response.status}`);
    }

    const result = await response.json();
    if (result.errors) {
      throw new Error(JSON.stringify(result.errors));
    }
    return result.data;
  }

  async login(username = DEFAULT_USERNAME, password = DEFAULT_PASSWORD) {
    const query = `
      mutation {
        login(username: "${username}", password: "${password}") {
          success
          token
          message
        }
      }
    `;
    const data = await this._execute(query);
    if (data.login.success) {
      this.token = data.login.token;
    }
    return data.login;
  }

  async register(username, firstName, lastName, password) {
    const query = `
      mutation {
        register(username: "${username}", firstName: "${firstName}", lastName: "${lastName}", password: "${password}") {
          success
          message
          userId
        }
      }
    `;
    return this._execute(query);
  }

  async me() {
    const query = `
      query Me {
        me {
          id
          username
          firstName
          lastName
          role
        }
      }
    `;
    return this._execute(query);
  }

  async books(search = '', categoryId = 0) {
    const searchStr = search ? `"${search}"` : '""';
    const query = `
      query Books {
        books(search: ${searchStr}, categoryId: ${categoryId}) {
          id
          title
          author
          price
          category
          description
        }
      }
    `;
    return this._execute(query);
  }

  async book(id) {
    const query = `
      query Book {
        book(id: ${id}) {
          id
          title
          author
          price
          category
          description
          stock
        }
      }
    `;
    return this._execute(query, { id });
  }

  async cart() {
    const query = `
      query Cart {
        cart {
          id
          items {
            bookId
            title
            quantity
            price
          }
          subtotal
          tax
          total
        }
      }
    `;
    return this._execute(query);
  }

  async addToCart(bookId, quantity = 1) {
    const query = `
      mutation {
        addToCart(bookId: ${bookId}, quantity: ${quantity}) {
          success
          message
        }
      }
    `;
    return this._execute(query);
  }

  async removeFromCart(bookId) {
    const query = `
      mutation {
        removeFromCart(bookId: ${bookId}) {
          success
          message
        }
      }
    `;
    return this._execute(query);
  }

  async orders() {
    const query = `
      query Orders {
        orders {
          id
          status
          totalAmount
          createdAt
          items {
            bookId
            title
            quantity
            price
          }
        }
      }
    `;
    return this._execute(query);
  }

  async createOrder() {
    const query = `
      mutation {
        createOrder {
          success
          orderId
          totalAmount
          message
        }
      }
    `;
    return this._execute(query);
  }

  async checkout(cardNumber, expiry, cvv) {
    const query = `
      mutation {
        checkout(cardNumber: "${cardNumber}", expiry: "${expiry}", cvv: "${cvv}") {
          success
          orderId
          message
        }
      }
    `;
    return this._execute(query);
  }

  async cancelOrder(orderId) {
    const query = `
      mutation {
        cancelOrder(orderId: "${orderId}") {
          success
          message
        }
      }
    `;
    return this._execute(query);
  }

  async createReview(bookId, rating, comment) {
    const query = `
      mutation {
        createReview(bookId: ${bookId}, rating: ${rating}, comment: "${comment}") {
          success
          reviewId
          message
        }
      }
    `;
    return this._execute(query);
  }

  async deleteReview(reviewId) {
    const query = `
      mutation {
        deleteReview(reviewId: ${reviewId}) {
          success
          message
        }
      }
    `;
    return this._execute(query);
  }

  async bookReviews(bookId) {
    const query = `
      query BookReviews {
        bookReviews(bookId: ${bookId}) {
          id
          userId
          rating
          comment
          createdAt
        }
      }
    `;
    return this._execute(query);
  }

  async myReviews() {
    const query = `
      query MyReviews {
        myReviews {
          id
          bookId
          rating
          comment
          createdAt
        }
      }
    `;
    return this._execute(query);
  }

  async updateProfile(firstName, lastName) {
    const query = `
      mutation {
        updateProfile(firstName: "${firstName}", lastName: "${lastName}") {
          success
          message
        }
      }
    `;
    return this._execute(query);
  }

  async searchBooks(query) {
    return this.books(query);
  }

  async applyCoupon(code) {
    const query = `
      mutation {
        applyCoupon(code: "${code}") {
          success
          message
          discount
        }
      }
    `;
    return this._execute(query);
  }

  async registerWebhook(url, events, secret = '') {
    const eventsStr = '["' + events.join('","') + '"]';
    const query = `
      mutation {
        registerWebhook(url: "${url}", events: ${eventsStr}, secret: "${secret}") {
          success
          webhookId
          message
        }
      }
    `;
    return this._execute(query);
  }

  async webhooks() {
    const query = `
      query Webhooks {
        webhooks {
          id
          url
          events
          createdAt
        }
      }
    `;
    return this._execute(query);
  }

  async testWebhook(webhookId) {
    const query = `
      mutation {
        testWebhook(webhookId: ${webhookId}) {
          success
          message
        }
      }
    `;
    return this._execute(query);
  }

  async logout() {
    const query = `
      mutation {
        logout {
          success
          message
        }
      }
    `;
    const data = await this._execute(query);
    this.token = null;
    return data.logout;
  }

  async adminStats() {
    const query = `
      query AdminStats {
        _adminStats {
          userCount
          bookCount
          totalRevenue
        }
      }
    `;
    return this._execute(query);
  }

  async adminOrders() {
    const query = `
      query AdminOrders {
        _adminAllOrders {
          id
          userId
          status
          totalAmount
          createdAt
        }
      }
    `;
    return this._execute(query);
  }

  async adminPayments() {
    const query = `
      query AdminPayments {
        _adminAllPayments {
          id
          orderId
          amount
          status
          createdAt
        }
      }
    `;
    return this._execute(query);
  }

  async proInventory() {
    const query = `
      query ProInventory {
        _proInventory {
          id
          title
          author
          price
          difficulty
          hint
        }
      }
    `;
    return this._execute(query);
  }
}

const client = new BookstoreMCPClient();

const tools = [
  {
    name: 'bookstore_login',
    description: 'Login to the Bookstore API with username and password',
    inputSchema: {
      type: 'object',
      properties: {
        username: { type: 'string', description: 'Username (default: admin)' },
        password: { type: 'string', description: 'Password (default: password123)' },
      },
    },
  },
  {
    name: 'bookstore_register',
    description: 'Register a new user account',
    inputSchema: {
      type: 'object',
      properties: {
        username: { type: 'string', description: 'Desired username' },
        firstName: { type: 'string', description: 'User first name' },
        lastName: { type: 'string', description: 'User last name' },
        password: { type: 'string', description: 'Desired password' },
      },
      required: ['username', 'firstName', 'lastName', 'password'],
    },
  },
  {
    name: 'bookstore_me',
    description: 'Get current authenticated user profile',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'bookstore_books',
    description: 'List all books with optional search and category filter',
    inputSchema: {
      type: 'object',
      properties: {
        search: { type: 'string', description: 'Search query for book title or author' },
        categoryId: { type: 'number', description: 'Filter by category ID (0 for all)' },
      },
    },
  },
  {
    name: 'bookstore_book',
    description: 'Get a specific book by ID',
    inputSchema: {
      type: 'object',
      properties: {
        id: { type: 'number', description: 'Book ID' },
      },
      required: ['id'],
    },
  },
  {
    name: 'bookstore_cart',
    description: 'Get current user shopping cart',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'bookstore_add_to_cart',
    description: 'Add a book to shopping cart',
    inputSchema: {
      type: 'object',
      properties: {
        bookId: { type: 'number', description: 'Book ID to add' },
        quantity: { type: 'number', description: 'Quantity (default: 1)' },
      },
      required: ['bookId'],
    },
  },
  {
    name: 'bookstore_remove_from_cart',
    description: 'Remove a book from shopping cart',
    inputSchema: {
      type: 'object',
      properties: {
        bookId: { type: 'number', description: 'Book ID to remove' },
      },
      required: ['bookId'],
    },
  },
  {
    name: 'bookstore_orders',
    description: 'Get current user order history',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'bookstore_create_order',
    description: 'Create an order from current cart',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'bookstore_checkout',
    description: 'Process checkout with payment',
    inputSchema: {
      type: 'object',
      properties: {
        cardNumber: { type: 'string', description: 'Credit card number (default: test card)' },
        expiry: { type: 'string', description: 'Expiry date (default: 12/25)' },
        cvv: { type: 'string', description: 'CVV (default: 123)' },
      },
    },
  },
  {
    name: 'bookstore_cancel_order',
    description: 'Cancel an order',
    inputSchema: {
      type: 'object',
      properties: {
        orderId: { type: 'string', description: 'Order ID to cancel' },
      },
      required: ['orderId'],
    },
  },
  {
    name: 'bookstore_create_review',
    description: 'Create a book review',
    inputSchema: {
      type: 'object',
      properties: {
        bookId: { type: 'number', description: 'Book ID to review' },
        rating: { type: 'number', description: 'Rating (1-5)' },
        comment: { type: 'string', description: 'Review comment' },
      },
      required: ['bookId', 'rating', 'comment'],
    },
  },
  {
    name: 'bookstore_delete_review',
    description: 'Delete a review',
    inputSchema: {
      type: 'object',
      properties: {
        reviewId: { type: 'number', description: 'Review ID to delete' },
      },
      required: ['reviewId'],
    },
  },
  {
    name: 'bookstore_book_reviews',
    description: 'Get reviews for a specific book',
    inputSchema: {
      type: 'object',
      properties: {
        bookId: { type: 'number', description: 'Book ID' },
      },
      required: ['bookId'],
    },
  },
  {
    name: 'bookstore_my_reviews',
    description: 'Get current user reviews',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'bookstore_update_profile',
    description: 'Update user profile',
    inputSchema: {
      type: 'object',
      properties: {
        firstName: { type: 'string', description: 'New first name' },
        lastName: { type: 'string', description: 'New last name' },
      },
    },
  },
  {
    name: 'bookstore_search',
    description: 'Search books by title or author',
    inputSchema: {
      type: 'object',
      properties: {
        query: { type: 'string', description: 'Search query' },
      },
      required: ['query'],
    },
  },
  {
    name: 'bookstore_apply_coupon',
    description: 'Apply coupon code to cart',
    inputSchema: {
      type: 'object',
      properties: {
        code: { type: 'string', description: 'Coupon code' },
      },
      required: ['code'],
    },
  },
  {
    name: 'bookstore_register_webhook',
    description: 'Register a webhook URL for events',
    inputSchema: {
      type: 'object',
      properties: {
        url: { type: 'string', description: 'Webhook URL' },
        events: { type: 'array', items: { type: 'string' }, description: 'Events to subscribe' },
        secret: { type: 'string', description: 'Webhook secret' },
      },
      required: ['url', 'events'],
    },
  },
  {
    name: 'bookstore_webhooks',
    description: 'List user webhooks',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'bookstore_test_webhook',
    description: 'Test a webhook',
    inputSchema: {
      type: 'object',
      properties: {
        webhookId: { type: 'number', description: 'Webhook ID to test' },
      },
      required: ['webhookId'],
    },
  },
  {
    name: 'bookstore_logout',
    description: 'Logout and clear token',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'bookstore_admin_stats',
    description: 'Get admin statistics',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'bookstore_admin_orders',
    description: 'Get all orders (admin)',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'bookstore_admin_payments',
    description: 'Get all payments (admin)',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'bookstore_pro_inventory',
    description: 'Get pro/hidden inventory (expert-level books)',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
];

async function handleToolCall(toolName, args) {
  try {
    switch (toolName) {
      case 'bookstore_login':
        return await client.login(args.username, args.password);
      case 'bookstore_register':
        return await client.register(args.username, args.firstName, args.lastName, args.password);
      case 'bookstore_me':
        return await client.me();
      case 'bookstore_books':
        return await client.books(args.search || '', args.categoryId || 0);
      case 'bookstore_book':
        return await client.book(args.id);
      case 'bookstore_cart':
        return await client.cart();
      case 'bookstore_add_to_cart':
        return await client.addToCart(args.bookId, args.quantity || 1);
      case 'bookstore_remove_from_cart':
        return await client.removeFromCart(args.bookId);
      case 'bookstore_orders':
        return await client.orders();
      case 'bookstore_create_order':
        return await client.createOrder();
      case 'bookstore_checkout':
        return await client.checkout(
          args.cardNumber || '4111111111111111',
          args.expiry || '12/25',
          args.cvv || '123'
        );
      case 'bookstore_cancel_order':
        return await client.cancelOrder(args.orderId);
      case 'bookstore_create_review':
        return await client.createReview(args.bookId, args.rating, args.comment);
      case 'bookstore_delete_review':
        return await client.deleteReview(args.reviewId);
      case 'bookstore_book_reviews':
        return await client.bookReviews(args.bookId);
      case 'bookstore_my_reviews':
        return await client.myReviews();
      case 'bookstore_update_profile':
        return await client.updateProfile(args.firstName, args.lastName);
      case 'bookstore_search':
        return await client.searchBooks(args.query);
      case 'bookstore_apply_coupon':
        return await client.applyCoupon(args.code);
      case 'bookstore_register_webhook':
        return await client.registerWebhook(args.url, args.events, args.secret || '');
      case 'bookstore_webhooks':
        return await client.webhooks();
      case 'bookstore_test_webhook':
        return await client.testWebhook(args.webhookId);
      case 'bookstore_logout':
        return await client.logout();
      case 'bookstore_admin_stats':
        return await client.adminStats();
      case 'bookstore_admin_orders':
        return await client.adminOrders();
      case 'bookstore_admin_payments':
        return await client.adminPayments();
      case 'bookstore_pro_inventory':
        return await client.proInventory();
      default:
        throw new Error(`Unknown tool: ${toolName}`);
    }
  } catch (error) {
    return { error: error.message };
  }
}

class BookstoreMCPServer {
  constructor() {
    this.server = new Server(
      {
        name: 'bookstore-mcp-server',
        version: '1.0.0',
      },
      {
        capabilities: {
          tools: {},
          resources: {},
        },
      }
    );

    this.server.setRequestHandler(ListToolsRequestSchema, async () => {
      return { tools };
    });

    this.server.setRequestHandler(CallToolRequestSchema, async (request) => {
      const { name, arguments: args } = request.params;
      const result = await handleToolCall(name, args);
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(result, null, 2),
          },
        ],
      };
    });

    this.server.setRequestHandler(ListResourcesRequestSchema, async () => {
      return { resources: [] };
    });

    this.server.setRequestHandler(ReadResourceRequestSchema, async (request) => {
      throw new Error('Resources not supported');
    });
  }

  async run() {
    const transport = new StdioServerTransport();
    await this.server.connect(transport);
    console.error('Bookstore MCP Server running on stdio');
  }
}

const server = new BookstoreMCPServer();
server.run().catch(console.error);

File: mcp/mcp_server.py
#!/usr/bin/env python3
"""
MCP Server for GraphQL Bookstore API
Allows AI/LLM models to interact with the Bookstore GraphQL API via MCP protocol.
"""

import os
import json
import requests
from typing import Any, Optional
from contextlib import contextmanager

API_URL = os.environ.get("API_URL", "http://localhost:4000/graphql")
DEFAULT_USERNAME = "admin"
DEFAULT_PASSWORD = "password123"


class BookstoreMCPClient:
    """Client for interacting with the GraphQL Bookstore API."""
    
    def __init__(self, api_url: str = API_URL):
        self.api_url = api_url
        self.token: Optional[str] = None
        self.session = requests.Session()
        self.session.headers.update({"Content-Type": "application/json"})
    
    def _execute(self, query: str, variables: Optional[dict] = None) -> dict:
        """Execute a GraphQL query."""
        body = {"query": query}
        if variables:
            body["variables"] = variables
        
        headers = {}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        
        response = self.session.post(
            self.api_url,
            json=body,
            headers=headers
        )
        response.raise_for_status()
        return response.json()
    
    def login(self, username: str, password: str) -> dict:
        """Login and store JWT token."""
        query = '''
        mutation Login($username: String!, $password: String!) {
            login(username: $username, password: $password) {
                success
                token
                message
            }
        }
        '''
        result = self._execute(query, {"username": username, "password": password})
        if result.get("data", {}).get("login", {}).get("success"):
            self.token = result["data"]["login"]["token"]
        return result
    
    def register(self, username: str, first_name: str, last_name: str, password: str) -> dict:
        """Register a new user."""
        query = '''
        mutation Register($username: String!, $firstName: String!, $lastName: String!, $password: String!) {
            register(username: $username, firstName: $firstName, lastName: $lastName, password: $password) {
                success
                message
                userId
            }
        }
        '''
        return self._execute(query, {
            "username": username,
            "firstName": first_name,
            "lastName": last_name,
            "password": password
        })
    
    def me(self) -> dict:
        """Get current authenticated user."""
        query = '''
        query Me {
            me {
                id
                username
                firstName
                lastName
                role
            }
        }
        '''
        return self._execute(query)
    
    def books(self, search: str = "", category_id: int = 0) -> dict:
        """List all books with optional search."""
        query = '''
        query Books($search: String, $categoryId: Int) {
            books(search: $search, categoryId: $categoryId) {
                id
                title
                author
                price
                category
                description
            }
        }
        '''
        return self._execute(query, {"search": search, "categoryId": category_id})
    
    def book(self, book_id: int) -> dict:
        """Get a specific book by ID."""
        query = '''
        query Book($id: Int!) {
            book(id: $id) {
                id
                title
                author
                price
                category
                description
                stock
            }
        }
        '''
        return self._execute(query, {"id": book_id})
    
    def cart(self) -> dict:
        """Get user's shopping cart."""
        query = '''
        query Cart {
            cart {
                id
                items {
                    bookId
                    title
                    quantity
                    price
                }
                subtotal
                tax
                total
            }
        }
        '''
        return self._execute(query)
    
    def add_to_cart(self, book_id: int, quantity: int = 1) -> dict:
        """Add item to shopping cart."""
        query = '''
        mutation AddToCart($bookId: Int!, $quantity: Int!) {
            addToCart(bookId: $bookId, quantity: $quantity) {
                success
                message
            }
        }
        '''
        return self._execute(query, {"bookId": book_id, "quantity": quantity})
    
    def remove_from_cart(self, book_id: int) -> dict:
        """Remove item from shopping cart."""
        query = '''
        mutation RemoveFromCart($bookId: Int!) {
            removeFromCart(bookId: $bookId) {
                success
                message
            }
        }
        '''
        return self._execute(query, {"bookId": book_id})
    
    def orders(self) -> dict:
        """Get user's orders."""
        query = '''
        query Orders {
            orders {
                id
                status
                totalAmount
                createdAt
                items {
                    bookId
                    title
                    quantity
                    price
                }
            }
        }
        '''
        return self._execute(query)
    
    def create_order(self) -> dict:
        """Create order from cart."""
        query = '''
        mutation CreateOrder {
            createOrder {
                success
                orderId
                totalAmount
                message
            }
        }
        '''
        return self._execute(query)
    
    def checkout(self, card_number: str, expiry: str, cvv: str) -> dict:
        """Process checkout with payment."""
        query = '''
        mutation Checkout($cardNumber: String!, $expiry: String!, $cvv: String!) {
            checkout(cardNumber: $cardNumber, expiry: $expiry, cvv: $cvv) {
                success
                orderId
                message
            }
        }
        '''
        return self._execute(query, {
            "cardNumber": card_number,
            "expiry": expiry,
            "cvv": cvv
        })
    
    def cancel_order(self, order_id: str) -> dict:
        """Cancel an order."""
        query = '''
        mutation CancelOrder($orderId: String!) {
            cancelOrder(orderId: $orderId) {
                success
                message
            }
        }
        '''
        return self._execute(query, {"orderId": order_id})
    
    def create_review(self, book_id: int, rating: int, comment: str) -> dict:
        """Create a book review."""
        query = '''
        mutation CreateReview($bookId: Int!, $rating: Int!, $comment: String!) {
            createReview(bookId: $bookId, rating: $rating, comment: $comment) {
                success
                reviewId
                message
            }
        }
        '''
        return self._execute(query, {
            "bookId": book_id,
            "rating": rating,
            "comment": comment
        })
    
    def delete_review(self, review_id: int) -> dict:
        """Delete a review."""
        query = '''
        mutation DeleteReview($reviewId: Int!) {
            deleteReview(reviewId: $reviewId) {
                success
                message
            }
        }
        '''
        return self._execute(query, {"reviewId": review_id})
    
    def book_reviews(self, book_id: int) -> dict:
        """Get reviews for a book."""
        query = '''
        query BookReviews($bookId: Int!) {
            bookReviews(bookId: $bookId) {
                id
                userId
                rating
                comment
                createdAt
            }
        }
        '''
        return self._execute(query, {"bookId": book_id})
    
    def my_reviews(self) -> dict:
        """Get current user's reviews."""
        query = '''
        query MyReviews {
            myReviews {
                id
                bookId
                rating
                comment
                createdAt
            }
        }
        '''
        return self._execute(query)
    
    def update_profile(self, first_name: str = None, last_name: str = None) -> dict:
        """Update user profile."""
        query = '''
        mutation UpdateProfile($firstName: String, $lastName: String) {
            updateProfile(firstName: $firstName, lastName: $lastName) {
                success
                message
            }
        }
        '''
        return self._execute(query, {
            "firstName": first_name,
            "lastName": last_name
        })
    
    def search_books(self, query: str) -> dict:
        """Search for books."""
        return self.books(search=query)
    
    def apply_coupon(self, code: str) -> dict:
        """Apply coupon code to cart."""
        query = '''
        mutation ApplyCoupon($code: String!) {
            applyCoupon(code: $code) {
                success
                message
                discount
            }
        }
        '''
        return self._execute(query, {"code": code})
    
    def register_webhook(self, url: str, events: list, secret: str = "") -> dict:
        """Register a webhook."""
        query = '''
        mutation RegisterWebhook($url: String!, $events: [String]!, $secret: String) {
            registerWebhook(url: $url, events: $events, secret: $secret) {
                success
                webhookId
                message
            }
        }
        '''
        return self._execute(query, {
            "url": url,
            "events": events,
            "secret": secret
        })
    
    def webhooks(self) -> dict:
        """Get user's webhooks."""
        query = '''
        query Webhooks {
            webhooks {
                id
                url
                events
                createdAt
            }
        }
        '''
        return self._execute(query)
    
    def test_webhook(self, webhook_id: int) -> dict:
        """Test a webhook."""
        query = '''
        mutation TestWebhook($webhookId: Int!) {
            testWebhook(webhookId: $webhookId) {
                success
                message
            }
        }
        '''
        return self._execute(query, {"webhookId": webhook_id})
    
    def admin_stats(self) -> dict:
        """Get admin statistics."""
        query = '''
        query AdminStats {
            _adminStats {
                userCount
                bookCount
                totalRevenue
            }
        }
        '''
        return self._execute(query)
    
    def admin_orders(self) -> dict:
        """Get all orders (admin)."""
        query = '''
        query AdminOrders {
            _adminAllOrders {
                id
                userId
                status
                totalAmount
                createdAt
            }
        }
        '''
        return self._execute(query)
    
    def admin_payments(self) -> dict:
        """Get all payments (admin)."""
        query = '''
        query AdminPayments {
            _adminAllPayments {
                id
                orderId
                amount
                status
                createdAt
            }
        }
        '''
        return self._execute(query)
    
    def pro_inventory(self) -> dict:
        """Get pro/hidden inventory."""
        query = '''
        query ProInventory {
            _proInventory {
                id
                title
                author
                price
                difficulty
                hint
            }
        }
        '''
        return self._execute(query)


_client: Optional[BookstoreMCPClient] = None


def get_client() -> BookstoreMCPClient:
    """Get or create the MCP client."""
    global _client
    if _client is None:
        _client = BookstoreMCPClient()
    return _client


def login(username: str = DEFAULT_USERNAME, password: str = DEFAULT_PASSWORD) -> dict:
    """Login to the bookstore API."""
    client = get_client()
    return client.login(username, password)


def logout() -> dict:
    """Logout (clears token locally)."""
    client = get_client()
    client.token = None
    return {"success": True, "message": "Logged out locally"}


def register(username: str, first_name: str, last_name: str, password: str) -> dict:
    """Register a new user."""
    client = get_client()
    return client.register(username, first_name, last_name, password)


def me() -> dict:
    """Get current authenticated user."""
    client = get_client()
    return client.me()


def books(search: str = "", category_id: int = 0) -> dict:
    """List all books with optional search."""
    client = get_client()
    return client.books(search, category_id)


def book(book_id: int) -> dict:
    """Get a specific book by ID."""
    client = get_client()
    return client.book(book_id)


def cart() -> dict:
    """Get user's shopping cart."""
    client = get_client()
    return client.cart()


def add_to_cart(book_id: int, quantity: int = 1) -> dict:
    """Add item to shopping cart."""
    client = get_client()
    return client.add_to_cart(book_id, quantity)


def remove_from_cart(book_id: int) -> dict:
    """Remove item from shopping cart."""
    client = get_client()
    return client.remove_from_cart(book_id)


def orders() -> dict:
    """Get user's orders."""
    client = get_client()
    return client.orders()


def create_order() -> dict:
    """Create order from cart."""
    client = get_client()
    return client.create_order()


def purchase_cart(card_number: str = "4111111111111111", expiry: str = "12/25", cvv: str = "123") -> dict:
    """Purchase cart with payment."""
    client = get_client()
    return client.purchase_cart(card_number, expiry, cvv)


def cancel_order(order_id: str) -> dict:
    """Cancel an order."""
    client = get_client()
    return client.cancel_order(order_id)


def create_review(book_id: int, rating: int, comment: str) -> dict:
    """Create a book review."""
    client = get_client()
    return client.create_review(book_id, rating, comment)


def delete_review(review_id: int) -> dict:
    """Delete a review."""
    client = get_client()
    return client.delete_review(review_id)


def book_reviews(book_id: int) -> dict:
    """Get reviews for a book."""
    client = get_client()
    return client.book_reviews(book_id)


def my_reviews() -> dict:
    """Get current user's reviews."""
    client = get_client()
    return client.my_reviews()


def update_profile(first_name: str = None, last_name: str = None) -> dict:
    """Update user profile."""
    client = get_client()
    return client.update_profile(first_name, last_name)


def search_books(query: str) -> dict:
    """Search for books by title or author."""
    client = get_client()
    return client.search_books(query)


def apply_coupon(code: str) -> dict:
    """Apply coupon code to cart."""
    client = get_client()
    return client.apply_coupon(code)


def register_webhook(url: str, events: list, secret: str = "") -> dict:
    """Register a webhook URL."""
    client = get_client()
    return client.register_webhook(url, events, secret)


def webhooks() -> dict:
    """Get user's webhooks."""
    client = get_client()
    return client.webhooks()


def test_webhook(webhook_id: int) -> dict:
    """Test a webhook."""
    client = get_client()
    return client.test_webhook(webhook_id)


def logout() -> dict:
    """Logout and clear token."""
    client = get_client()
    return client.logout()


def admin_stats() -> dict:
    """Get admin statistics."""
    client = get_client()
    return client.admin_stats()


def admin_orders() -> dict:
    """Get all orders (admin)."""
    client = get_client()
    return client.admin_orders()


def admin_payments() -> dict:
    """Get all payments (admin)."""
    client = get_client()
    return client.admin_payments()


def pro_inventory() -> dict:
    """Get pro/hidden inventory (expert-level books)."""
    client = get_client()
    return client.pro_inventory()


if __name__ == "__main__":
    import sys
    
    client = get_client()
    print("Bookstore MCP Client")
    print("=" * 40)
    print(f"API URL: {client.api_url}")
    print("\nLogging in with default credentials...")
    
    result = client.login(DEFAULT_USERNAME, DEFAULT_PASSWORD)
    if result.get("data", {}).get("login", {}).get("success"):
        print("Logged in successfully!")
        me = client.me()
        print(f"User: {me.get('data', {}).get('me', {}).get('username')}")
        
        books = client.books()
        book_list = books.get("data", {}).get("books", [])
        print(f"Available books: {len(book_list)}")
    else:
        print("Login failed:", result)
        sys.exit(1)
